GeneratorInterface/GenFilters/src/PythiaFilterGammaGamma.cc

```diff
@@ -251,56 +161,48 @@ bool PythiaFilterGammaGamma::filter(edm::Event& iEvent, const edm::EventSetup& i
 
 	if(acceptPrompts) {
 	  bool isPrompt=false;
-	  if((*itSeed)->status() == 1&&(*itSeed)->pdg_id() == 22) {
-	    const GenParticle* mother = (*itSeed)->production_vertex() ?
-	      *((*itSeed)->production_vertex()->particles_in_const_begin()) : 0;
-	    if(mother) {
-	      if(mother->pdg_id()>=-22&&mother->pdg_id()<=22) {
-		const GenParticle* motherMother = (mother != 0  && mother->production_vertex()) ?
-		  *(mother->production_vertex()->particles_in_const_begin()) : 0;
-		if(motherMother) {
-		  if(motherMother->pdg_id()>=-22&&motherMother->pdg_id()<=22) {
-		    if((*itSeed)->momentum().perp()>promptPtThreshold) {
-		      isPrompt=true;
-		    }
-		  }
-		}
-	      }
-	    }
-	  }
+       
+	  const GenParticle* mother = (*itSeed)->production_vertex() ?
+	    *((*itSeed)->production_vertex()->particles_in_const_begin()) : 0;
+	  const GenParticle* motherMother = (mother != 0 && mother->production_vertex()) ?
+	    *(mother->production_vertex()->particles_in_const_begin()) : 0;
+	  const GenParticle* motherMotherMother = (motherMother != 0 && motherMother->production_vertex()) ?
+	    *(motherMother->production_vertex()->particles_in_const_begin()) : 0;
+
+
+	  if ( ((mother->pdg_id() != (*itSeed)->pdg_id()) && (mother->pdg_id()<-22||mother->pdg_id()>22) && mother->status() == 2) ||
+	       ((motherMother->pdg_id() != (*itSeed)->pdg_id()) && (motherMother->pdg_id()<-22||motherMother->pdg_id()>22) && motherMother->status() == 2) ||
+	       ((motherMotherMother->pdg_id() != (*itSeed)->pdg_id()) && (motherMotherMother->pdg_id()<-22||motherMotherMother->pdg_id()>22) && motherMotherMother->status() == 2) ) isPrompt=false;
```

Comment:
For checking if mother is a hadron, better to use std::abs(pdg_id)>100  (to avoid some other BSM particles and special states)

Also would be better here to have a while loop, rather than a hardcoded number of mothers to check.  (Loop can stop at first mother with pdg id different from tested photon/electron, or else break when no mother is found)